man/man4/mmv.4
'\"! tbl | nroff \-man
'\"macro stdmacro
.\"
.\" Copyright (c) 2009 Max Matveev
.\" Copyright (c) 2009 Aconex.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\" 
.\"
.TH MMV 4 "" "Performance Co-Pilot"
.SH NAME
\f3mmv\f1 \- Memory Mapped Values for Performance Co-Pilot
.SH SYNOPSIS
.I $PCP_TMP_DIR/mmv/<file>
.SH DESCRIPTION
The files in \f2$PCP_TMP_DIR/mmv\f1 are generated by 
\f2mmv_stats_init\f1() function from \f3libpcp_mmv\f1 library. There could
be multiple files in this directory, each file representing a single source
of the performance metrics. The metrics are harvested by the \f2mmv\f1 PMDA
which exports them to the rest of the Performance Co-Pilot infrastructure.
.SH FILE FORMAT
Each file starts with the following header:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Name
_
0	4	tag == "MMV\\0"
_
4	4	Version
_
8	8	Generation 1
_
16	8	Generation 2
_
24	4	Number of TOC entries
_
28	4	Flags
_
32	4	Process identifier (PID)
_
36	4	Cluster identifier
.TE
.PP
.PP
The generation numbers are timestamps at the time of file
creation, and must match for the file to be considered by
the MMV PMDA.
.PP
The flags can specify ways in which the client would like
the MMV PMDA to behave - e.g. the MMV_FLAG_PROCESS flag
specifies that only if the process identified by PID is
currently running should those values be exported.
.PP
Finally, if set, the cluster identifier is a hint to the MMV
PMDA as to what cluster should be used with this application
when forming the individual metric identifiers.
A performance metric identifier (see \f2PMDA\f1(3)) consists of
the PMDA domain number, the cluster number, and the individual
item numbers described in the Metrics section.
.PP
The header is followed by at least 2 TOC sections:
one section for metrics and another for values.
The TOC section has the following format:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Value
_
0	4	Section Type (see \f2mmv_stats.h\f1)
_
4	4	Number of entries in the section
_
8	8	Section's offset from the start of the file
.TE
.PP
.PP
The section types are:
.IP
1:
Indoms (instance domain definitions)
.IP
2:
Instances
.IP
3:
Metrics (metric definitions)
.IP
4:
Values
.IP
5:
String
.PP
The only mandatory sections are Metrics and Values.
Indoms and Instances sections only appear if there are metrics with
multiple instances.
String sections only appear if there are metrics with string values,
or when Metrics or Indoms are defined with help text.
.PP
The entries in the Indoms section have the following format:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Value
_
0	4	Unique serial number for this domain
_
4	4	Number of entries in the domain
_
8	8	Offset to first instance
_
16	8	Short help text offset
_
24	8	Long help text offset
.TE
.PP
.PP
The entries in the Instances section have the following format:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Value
_
0	8	Offset into the indom section
_
8	4	Unused padding (zero filled)
_
12	4	Internal instance identifier
_
16	64	External instance identifier
.TE
.PP
.PP
The entries in the Metrics section have the following format:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Value
_
0	64	Metric Name
_
64	4	Metric Item (see \f2PMDA\f1(3))
_
68	4	Metric Type (see \f2mmv_stats.h\f1)
_
72	4	Semantics (see \f2PMAPI\f1(3))
_
76	4	Dimensions (see \f2PMAPI\f1(3))
_
80	4	Instance Domain ID
_
84	4	Unused padding (zero filled)
_
88	8	Short help text offset
_
96	8	Long help text offset
.TE
.PP
.PP
The entries in the Values section have the following format:
.TS
box,center;
c | c | c
n | n | l.
Offset	Length	Value
_
0	8	\f3pmAtomValue\f1 (see \f2PMAPI\f1(3))
_
8	8	Extra space for STRING and ELAPSED
_
16	8	Offset into the Metrics section
_
24	8	Offset into the Instances section
.TE
.PP
.PP
Each entry in the strings section is a 256 byte character array,
containing a single NULL-terminated character string.
So each string has a maximum length of 256 bytes, which includes
the terminating NULL.
.PP
.SH SEE ALSO
.BR PCPIntro (1),
.BR PMAPI (3),
.BR mmv_stats_init (3),
.BR pcp.conf (4)
and
.BR pcp.env (4).


CHANGELOG
pcp-2.8.11 (6 July 2009)
    - Rework namespace file generation for Perl PMDAs.

pcp-2.8.10 (1 July 2009)
    - RPM packaging fixes for Perl modules.

pcp-2.8.9 (1 July 2009)
    - Remove Cygwin support, we're now committed to native Win32
      port - first production install yesterday, hooray!
    - Further Win32 work - wrapper batch files for shell scripts
      run as commands, moved daemon configs into $PCP_DIR/etc,
      bug fixes to process creation code.
    - Significant work on the MMV PMDA, including revamped API,
      Perl API, updated on-disk support with string values and
      help text support, amongst other changes.
    - Fix build of Perl PMDA module when PCP not installed.
    - Fix Zimbra PMDA status metrics.

pcp-2.8.8 (9 June 2009)
    - Added a Zimbra Collaboration Suite PMDA.
    - Build fixes for OpenSolaris.
    - Perl PMDA fixes in tail mode.
    - Correct physical memory reporting in Windows PMDA on 32 bit
      machines.
    - Other small tweaks and improvements to Windows PMDA also.

pcp-2.8.7 (29 May 2009)
    - Windows PMDA major rework to improve memory footprint.
    - Yet another Mac OS X build issue resolution.

pcp-2.8.6 (26 May 2009)
    - Fix build on Mac OS X after lex warning cleanups.
    - Fix pmdate build when PCP headers not in the root.
    - A single PMNS domain number file now used in-tree.
    - Added tmpfs filesystem metrics to Linux PMDA.

pcp-2.8.5 (21 May 2009)
    - Packaging tweaks for Debian (builddefs/rules moved).
    - Make timezone manipulation in Windows work (mimic MSYS).
    - Numerous warnings fixed.
    - Several changes to tempfile handling for Vista.
    - Make setting pcp_rc_dir in configure.in more robust.
    - Further pathname separator auditing for Windows.
    - Fixed a libpcp_gui linker issue on Mac OS X.

pcp-2.8.4 (12 May 2009)
    - Descend into src/bashrc.  D'oh!

pcp-2.8.3 (11 May 2009)
    - Bash auto-completion of metric names.
    - Fix a memory leak in xfs project quotas metrics.
    - Several Win32 fixes for Windows Server 2008.
    - Add Linux per-CPU and aggregate "guest" CPU utilisation metrics.
    - Add a pmlogsummary option to report a "header" line (-H).
    - Add optional pmstat support for time control via pmtime.
    - Add pmcd.client.* metrics for identifying connected clients.

pcp-2.8.2 (24 April 2009)
    - Fix pmie bug in the handling of && and || operators.
    - Improve daily pmie/pmlogger script behaviour.

pcp-2.8.1 (19 April 2009)
    - Fix build issue with libpcp_mmv - would not build-from-source
      if <pcp/pmapi.h> was not in the root filesystem.
    - Finer control of pmie2col(1) reporting precision.

pcp-2.8.0 (9 April 2009)
    - Linux netfilter (IP connection tracking) PMDA included
    - Memory Mapped Value (MMV) PMDA and client library included
    - Linux bonding (bonded network interface) PMDA included
    - Lustre PMDA included
    - Added pcp_gui library as pcpmon replacement, supporting the
      new (open source) version of pmtime for console tools.
    - Perl PCP::LogSummary module included.
    - KVM (Linux Kernel Virtualisation layer) PMDA
    - Infiniband PMDA included
    - VMware PMDA included
    - Perl PCP::PMDA module completed, stabilised and incorporated.
    - Default pmcd and pmproxy ports changed to IANA registered ones.
      ( The old behaviour - i.e. both original and IANA ports - is
      still available using PMCD_PORTS and PMPROXY_PORTS variables ).

pcp-2.7.7 (5 September 2008)
    - Infiniband metric enhancements
    - Handle missing "which" binary during rpm installation
    - Added new quota metrics (XFS project quota)
    - Aggregate and per-CPU hypervisor "steal" time metrics added
    - NFS v4 metrics
    - commitLimit meminfo metric added
    - Namespace locking fix in pmdaproc.sh
    - Additional Windows memory metrics
    - Several pmie fixes from Ken
    - Kens pmlogreduce archive corruption fix
    - Kens pmproxy PDU size fix
    - Nathan's pmlogsummary sum calulation fix
    - Added a (cheap) runnable processes metric.

pcp-2.7.4 (7 September 2007)
    - MaxOSX and Debian/Ubuntu build/package fixes
    - Add sqlserver active_transactions metrics to the Windows PMDA.
    - Fix a pmie_check typo causing mis-identification of pmie processes.
    - Allow pmie and/or operators to function with some data missing.
    - Resolve path naming issues with more recent versions of autoconf.
    - pmlogsummary report sum option
    - pmval kmtime support
    - Additional Linux SNMP metrics
    - Fix Linux vmstat nr_slab metrics
    - Make pcp status command report build version
    - Windows split_io metrics
    - Fix pmdapmcd empty pmie instance
    - pmdamailq filename regex
    - Windows TCP metrics
    - pmie log file rotation
    - Changed the default compression program to be bzip2(1)

pcp-2.7.3 (July 2007)
    - add network.ib.control to timeout infiniband stats workthread
    - add mem.util.anonpages 

pcp-2.7.2 (8 Jun 2007)
    - pcp doc updates
    - pmdalinux death from open file descriptors: pclose needed in network.ib
    - network.ib stats updated to cope with OFED 1.2 changes
    - some numa.link fixes for shub2 & NL4
    - Improve start/stop times of pmcd,pmlogger,pmie
    - create portable pmsleep (subsecond sleep) exe

pcp-2.7.0 (7 Feb 2007)
    This log has been allowed to lapse for some time. Blanket catch-up..
    the following PVs describe changes made in that time:
    - 947510 - UNIX95 patches break pcp tools
    - 948548 - [SUSE#182852] Buffer overflow in linux proc_pmda
    - 948551 - Update pcp configure and build infrastructure
    - 948799 - Move telnet-probe to oss part of pcp
    - 948958 - *nodeid conversions have inconsistent naming
    - 952623 - possible use-after-free of pmProfile objects used by pmda
    - 952932 - pcp-open RPM requires libpcp.so.2 but doesn't provide it
    - 953015 - Promote libpcp_pmc to DSO
    - 953301 - Update irix pcp bits

src/cpan/MMV/MMV.pm
package PCP::MMV;

use strict;
use warnings;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;
require DynaLoader;

@ISA = qw(Exporter DynaLoader);
@EXPORT = qw(
    mmv_stats_init mmv_units
	mmv_lookup_value_desc
	mmv_inc_value mmv_set_value mmv_set_string
	mmv_stats_add mmv_stats_inc
	mmv_stats_add_fallback mmv_stats_inc_fallback
	mmv_stats_interval_start mmv_stats_interval_end
	mmv_stats_set_string
    MMV_FLAG_NOPREFIX MMV_FLAG_PROCESS
    MMV_INDOM_NULL
    MMV_TYPE_NOSUPPORT
    MMV_TYPE_I32 MMV_TYPE_U32
    MMV_TYPE_I64 MMV_TYPE_U64
    MMV_TYPE_FLOAT MMV_TYPE_DOUBLE
    MMV_TYPE_STRING MMV_TYPE_ELAPSED
    MMV_COUNT_ONE
    MMV_SEM_COUNTER MMV_SEM_INSTANT MMV_SEM_DISCRETE
    MMV_SPACE_BYTE MMV_SPACE_KBYTE MMV_SPACE_MBYTE
    MMV_SPACE_GBYTE MMV_SPACE_TBYTE
    MMV_TIME_NSEC MMV_TIME_USEC MMV_TIME_MSEC
    MMV_TIME_SEC MMV_TIME_MIN MMV_TIME_HOUR
);
@EXPORT_OK = qw();
$VERSION = '0.01';

sub MMV_INDOM_NULL	{ 0xffffffff; }

# flags for pmdammv
sub MMV_FLAG_NOPREFIX	{ 0x1; } # metric names not prefixed by file name
sub MMV_FLAG_PROCESS	{ 0x2; } # instrumented process must be running

# data type of metric values
sub MMV_TYPE_NOSUPPORT	{ 0xffffffff; }	# not implemented in this version
sub MMV_TYPE_I32	{ 0; }	# 32-bit signed integer
sub MMV_TYPE_U32	{ 1; }	# 32-bit unsigned integer
sub MMV_TYPE_I64	{ 2; }	# 64-bit signed integer
sub MMV_TYPE_U64	{ 3; }	# 64-bit signed integer
sub MMV_TYPE_FLOAT	{ 4; }	# 32-bit floating point
sub MMV_TYPE_DOUBLE	{ 5; }	# 64-bit floating point
sub MMV_TYPE_STRING	{ 6; }	# null-terminated string
sub MMV_TYPE_ELAPSED	{ 10; }	# 64-bit elapsed time

# units - space scale
sub MMV_SPACE_BYTE	{ 0; }  # bytes
sub MMV_SPACE_KBYTE	{ 1; }  # kilobytes
sub MMV_SPACE_MBYTE	{ 2; }  # megabytes
sub MMV_SPACE_GBYTE	{ 3; }  # gigabytes
sub MMV_SPACE_TBYTE	{ 4; }  # terabytes

# units - time scale
sub MMV_TIME_NSEC	{ 0; }  # nanoseconds
sub MMV_TIME_USEC	{ 1; }  # microseconds
sub MMV_TIME_MSEC	{ 2; }  # milliseconds
sub MMV_TIME_SEC	{ 3; }  # seconds
sub MMV_TIME_MIN	{ 4; }  # minutes
sub MMV_TIME_HOUR	{ 5; }  # hours

# units - count scale   (for metrics such as count events, syscalls,
# interrupts, etc - these are simply powers of ten and not enumerated here
# (e.g. 6 for 10^6, or -3 for 10^-3).
sub MMV_COUNT_ONE	{ 0; }  # 1

# semantics/interpretation of metric values
sub MMV_SEM_COUNTER	{ 1; }	# cumulative counter, monotonic increasing
sub MMV_SEM_INSTANT	{ 3; }	# instantaneous value, continuous domain
sub MMV_SEM_DISCRETE	{ 4; }	# instantaneous value, discrete domain


bootstrap PCP::MMV $VERSION;

1;
__END__

=head1 NAME

PCP::MMV - Perl module for Memory Mapped Value instrumentation

=head1 SYNOPSIS

  use PCP::MMV;

=head1 DESCRIPTION

The PCP::MMV Perl module contains the language bindings for
building Perl programs instrumented with the Performance Co-Pilot
Memory Mapped Value infrastructure - an efficient data transport
mechanism for making performance data from within a Perl program
available as PCP metrics using the MMV PMDA.

=head1 SEE ALSO

mmv_stats_init(3), mmv(4) and pmda(3).

The PCP mailing list pcp@oss.sgi.com can be used for questions about
this module.

Further details can be found at http://oss.sgi.com/projects/pcp

=head1 AUTHOR

Nathan Scott, E<lt>nathans@debian.orgE<gt>

Copyright (C) 2009 by Aconex.

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License, version 2 (see
the "COPYING" file in the PCP source tree for further details).

=cut
